File: Assignment03/Assignment03/parser.py
# CS 414 - Assignment 03, Step 3
# Recursive descent parser for the simple shell command language.
# No parser generator used - just one function per grammar rule below.
#
# Grammar (EBNF):
#
#   command     = ls_cmd | cd_cmd | cat_cmd | print_cmd | exec_cmd ;
#   ls_cmd      = "ls" , [ path ] ;
#   cd_cmd      = "cd" , [ path ] ;
#   cat_cmd     = "cat" , FILENAME ;
#   print_cmd   = "print" , FILENAME ;
#   exec_cmd    = "exec" , FILENAME ;
#   path        = "/" , [ folder_list ] | folder_list ;
#   folder_list = FOLDERNAME , { "/" , FOLDERNAME } ;
#
# Start symbol: command
#
# Reserved words: ls, cd, cat, print, exec (can't also be used as a
# folder/file name - the tokenizer just checks for these first)

from tokenizer import tokenize, LS, CD, CAT, PRINT, EXEC, FILENAME, FOLDERNAME, SLASH, EOF


class ParseError(Exception):
    pass


class Path:
    def __init__(self, absolute, folders):
        self.absolute = absolute
        self.folders = folders

    def __repr__(self):
        kind = "absolute" if self.absolute else "relative"
        return f"Path({kind}, {self.folders})"


class Command:
    def __init__(self, name, path=None, filename=None):
        self.name = name
        self.path = path
        self.filename = filename

    def __repr__(self):
        if self.name in (LS, CD):
            return f"{self.name}(path={self.path})"
        return f"{self.name}(filename={self.filename!r})"


class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0

    def peek(self):
        return self.tokens[self.pos]

    def advance(self):
        tok = self.tokens[self.pos]
        if tok.type != EOF:
            self.pos += 1
        return tok

    def expect(self, type):
        tok = self.peek()
        if tok.type != type:
            raise ParseError(f"expected {type} but got {tok.type} ({tok.value!r})")
        return self.advance()

    # command = ls_cmd | cd_cmd | cat_cmd | print_cmd | exec_cmd ;
    def parse_command(self):
        tok = self.peek()

        if tok.type in (LS, CD):
            self.advance()
            path = self.parse_optional_path()
            self.expect(EOF)
            return Command(tok.type, path=path)

        if tok.type in (CAT, PRINT, EXEC):
            self.advance()
            fname = self.expect(FILENAME)
            self.expect(EOF)
            return Command(tok.type, filename=fname.value)

        raise ParseError(f"expected a command but got {tok.type} ({tok.value!r})")

    # [ path ]
    def parse_optional_path(self):
        if self.peek().type in (SLASH, FOLDERNAME):
            return self.parse_path()
        return None

    # path = "/" , [ folder_list ] | folder_list ;
    def parse_path(self):
        if self.peek().type == SLASH:
            self.advance()
            folders = self.parse_folder_list() if self.peek().type == FOLDERNAME else []
            return Path(True, folders)
        return Path(False, self.parse_folder_list())

    # folder_list = FOLDERNAME , { "/" , FOLDERNAME } ;
    def parse_folder_list(self):
        folders = [self.expect(FOLDERNAME).value]
        while self.peek().type == SLASH:
            self.advance()
            folders.append(self.expect(FOLDERNAME).value)
        return folders


def parse(text):
    tokens = tokenize(text)
    return Parser(tokens).parse_command()


if __name__ == "__main__":
    import sys
    line = " ".join(sys.argv[1:]) or "cd docs/reports"
    try:
        print(parse(line))
    except Exception as e:
        print("ERROR:", e)

File: Assignment05/parser.cpp
#include "parser.h"

#include <sstream>

namespace shell {

ParseError::ParseError(const std::string& message) : std::runtime_error(message) {}

std::string PathNode::toString() const {
    std::ostringstream oss;
    oss << "Path(" << (isAbsolute ? "absolute" : "relative") << ", [";
    for (size_t i = 0; i < folders.size(); ++i) {
        if (i) oss << ", ";
        oss << folders[i];
    }
    oss << "])";
    return oss.str();
}

std::string ExprNode::toString() const {
    if (isVariable) return variable;
    std::ostringstream oss;
    oss << "(" << left->toString() << " " << op << " " << right->toString() << ")";
    return oss.str();
}

static const char* kindName(CommandKind k) {
    switch (k) {
        case CommandKind::LS: return "Ls";
        case CommandKind::CD: return "Cd";
        case CommandKind::CAT: return "Cat";
        case CommandKind::PRINT: return "Print";
        case CommandKind::EXEC: return "Exec";
        case CommandKind::SET: return "Set";
        case CommandKind::ECHO: return "Echo";
    }
    return "?";
}

std::string Command::toString() const {
    std::ostringstream oss;
    oss << kindName(kind) << "(";
    if (kind == CommandKind::LS || kind == CommandKind::CD) {
        oss << "path=" << (path.has_value() ? path->toString() : "None");
    } else if (kind == CommandKind::SET) {
        oss << "var='" << filename << "', expr=" << (expr ? expr->toString() : "?");
    } else if (kind == CommandKind::ECHO) {
        oss << "var='" << filename << "'";
    } else {
        oss << "filename='" << filename << "'";
    }
    oss << ")";
    return oss.str();
}

Parser::Parser(std::vector<Token> tokens, SymbolTable& symbols)
    : tokens_(std::move(tokens)), symbols_(symbols) {}

const Token& Parser::peek() const { return tokens_[pos_]; }

Token Parser::advance() {
    Token tok = tokens_[pos_];
    if (tok.type != TokenType::END) {
        ++pos_;
    }
    return tok;
}

Token Parser::expect(TokenType type) {
    const Token& tok = peek();
    if (tok.type != type) {
        std::ostringstream oss;
        oss << "expected " << tokenTypeName(type) << " but found " << tokenTypeName(tok.type)
            << " ('" << tok.lexeme << "') at position " << tok.pos;
        throw ParseError(oss.str());
    }
    return advance();
}

// Variables are "defined upon first use" - just make sure it has an entry.
void Parser::noteVariable(const std::string& name) {
    symbols_.emplace(name, "");
}

// command = ls_cmd | cd_cmd | cat_cmd | print_cmd | exec_cmd | set_cmd | echo_cmd ;
Command Parser::parseCommand() {
    const Token& tok = peek();
    Command node;

    switch (tok.type) {
        case TokenType::LS: node = lsCmd(); break;
        case TokenType::CD: node = cdCmd(); break;
        case TokenType::CAT: node = catCmd(); break;
        case TokenType::PRINT: node = printCmd(); break;
        case TokenType::EXEC: node = execCmd(); break;
        case TokenType::SET: node = setCmd(); break;
        case TokenType::ECHO: node = echoCmd(); break;
        default: {
            std::ostringstream oss;
            oss << "expected a command (ls, cd, cat, print, exec, set, echo) but found "
                << tokenTypeName(tok.type) << " ('" << tok.lexeme << "') at position " << tok.pos;
            throw ParseError(oss.str());
        }
    }

    // A well-formed command consumes every token; anything left over is an error.
    expect(TokenType::END);
    return node;
}

Command Parser::lsCmd() {
    expect(TokenType::LS);
    Command c;
    c.kind = CommandKind::LS;
    c.path = optionalPath();
    return c;
}

Command Parser::cdCmd() {
    expect(TokenType::CD);
    Command c;
    c.kind = CommandKind::CD;
    c.path = optionalPath();
    return c;
}

Command Parser::catCmd() {
    expect(TokenType::CAT);
    Token f = fileOrVar();
    Command c;
    c.kind = CommandKind::CAT;
    c.filename = f.lexeme;
    return c;
}

Command Parser::printCmd() {
    expect(TokenType::PRINT);
    Token f = fileOrVar();
    Command c;
    c.kind = CommandKind::PRINT;
    c.filename = f.lexeme;
    return c;
}

Command Parser::execCmd() {
    expect(TokenType::EXEC);
    Token f = fileOrVar();
    Command c;
    c.kind = CommandKind::EXEC;
    c.filename = f.lexeme;
    return c;
}

// set_cmd = "set" , VARIABLE , "=" , expr ;
Command Parser::setCmd() {
    expect(TokenType::SET);
    Token v = expect(TokenType::VARIABLE);
    noteVariable(v.lexeme);
    expect(TokenType::ASSIGN);
    auto e = expr();

    // We're only parsing, not evaluating, so we store the expression's
    // text as the variable's "current value" (values are strings anyway).
    symbols_[v.lexeme] = e->toString();

    Command c;
    c.kind = CommandKind::SET;
    c.filename = v.lexeme;
    c.expr = e;
    return c;
}

// echo_cmd = "echo" , VARIABLE ;
Command Parser::echoCmd() {
    expect(TokenType::ECHO);
    Token v = expect(TokenType::VARIABLE);
    noteVariable(v.lexeme);
    Command c;
    c.kind = CommandKind::ECHO;
    c.filename = v.lexeme;
    return c;
}

// [ path ]
std::optional<PathNode> Parser::optionalPath() {
    TokenType t = peek().type;
    if (t == TokenType::SLASH || t == TokenType::FOLDERNAME || t == TokenType::VARIABLE) {
        return path();
    }
    return std::nullopt;
}

// path = "/" , [ folder_list ] | folder_list ;
PathNode Parser::path() {
    if (peek().type == TokenType::SLASH) {
        advance();
        std::vector<std::string> folders;
        if (peek().type == TokenType::FOLDERNAME || peek().type == TokenType::VARIABLE) {
            folders = folderList();
        }
        return PathNode{true, folders};
    }
    return PathNode{false, folderList()};
}

// folder_list = folder_or_var , { "/" , folder_or_var } ;
std::vector<std::string> Parser::folderList() {
    std::vector<std::string> folders;
    folders.push_back(folderOrVar().lexeme);
    while (peek().type == TokenType::SLASH) {
        advance();
        folders.push_back(folderOrVar().lexeme);
    }
    return folders;
}

// folder_or_var = FOLDERNAME | VARIABLE ;
Token Parser::folderOrVar() {
    if (peek().type == TokenType::VARIABLE) {
        Token t = advance();
        noteVariable(t.lexeme);
        return t;
    }
    return expect(TokenType::FOLDERNAME);
}

// file_or_var = FILENAME | VARIABLE ;
Token Parser::fileOrVar() {
    if (peek().type == TokenType::VARIABLE) {
        Token t = advance();
        noteVariable(t.lexeme);
        return t;
    }
    return expect(TokenType::FILENAME);
}

// expr = term , { ( "+" | "-" ) , term } ;
std::shared_ptr<ExprNode> Parser::expr() {
    auto node = term();
    while (peek().type == TokenType::PLUS || peek().type == TokenType::MINUS) {
        Token opTok = advance();
        auto parent = std::make_shared<ExprNode>();
        parent->isVariable = false;
        parent->op = (opTok.type == TokenType::PLUS) ? '+' : '-';
        parent->left = node;
        parent->right = term();
        node = parent;
    }
    return node;
}

// term = factor , { ( "*" | "/" ) , factor } ;
std::shared_ptr<ExprNode> Parser::term() {
    auto node = factor();
    while (peek().type == TokenType::STAR || peek().type == TokenType::SLASH) {
        Token opTok = advance();
        auto parent = std::make_shared<ExprNode>();
        parent->isVariable = false;
        parent->op = (opTok.type == TokenType::STAR) ? '*' : '/';
        parent->left = node;
        parent->right = factor();
        node = parent;
    }
    return node;
}

// factor = VARIABLE | "(" , expr , ")" ;
std::shared_ptr<ExprNode> Parser::factor() {
    if (peek().type == TokenType::VARIABLE) {
        Token v = advance();
        noteVariable(v.lexeme);
        auto node = std::make_shared<ExprNode>();
        node->isVariable = true;
        node->variable = v.lexeme;
        return node;
    }
    if (peek().type == TokenType::LPAREN) {
        advance();
        auto node = expr();
        expect(TokenType::RPAREN);
        return node;
    }

    std::ostringstream oss;
    oss << "expected a variable or '(' but found " << tokenTypeName(peek().type)
        << " ('" << peek().lexeme << "') at position " << peek().pos;
    throw ParseError(oss.str());
}

Command parse(const std::string& text, SymbolTable& symbols) {
    std::vector<Token> tokens = tokenize(text);
    Parser p(std::move(tokens), symbols);
    return p.parseCommand();
}

Command parse(const std::string& text) {
    SymbolTable symbols;
    return parse(text, symbols);
}

}  // namespace shell
